crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1726 @@
+
+    /* All done with |y|, now free to reuse tmp[0] for FIPS 203 |e1| */
+    if (!gencbd_vector(e1, cbd_2, &counter, r, rank, mdctx, key))
+        return 0;
+    vector_add(u, e1, rank);
+    vector_compress(u, du, rank);
+    vector_encode(out, u, du, rank);
+
+    /* All done with |e1|, now free to reuse tmp[0] for FIPS 203 |e2| */
+    memcpy(input, r, ML_KEM_RANDOM_BYTES);
+    input[ML_KEM_RANDOM_BYTES] = counter;
+    if (!cbd_2(e2, input, mdctx, key))
+        return 0;
+    scalar_add(&v, e2);
+
+    /* All Done with |e2|, now free to reuse tmp[0] for FIPS 203 |mu| */
+    scalar_decode_1(mu, message);
+    scalar_decompress(mu, 1);
+    scalar_add(&v, mu);
+    scalar_compress(&v, dv);
+    scalar_encode(out + vinfo->u_vector_bytes, &v, dv);
+    return 1;
+}
+
+/*
+ * FIPS 203, Section 5.3, Algorithm 15: K-PKE.Decrypt.
+ */
+static void
+decrypt_cpa(uint8_t *out, const uint8_t *ctext, scalar *u,
+            const ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+    scalar v, mask;
+    int rank = vinfo->rank;
+    int du = vinfo->du;
+    int dv = vinfo->dv;
+
+    vector_decode(u, ctext, du, rank);
+    vector_decompress(u, du, rank);
+    vector_ntt(u, rank);
+    scalar_decode(&v, ctext + vinfo->u_vector_bytes, dv);
+    scalar_decompress(&v, dv);
+    inner_product(&mask, key->s, u, rank);
+    scalar_inverse_ntt(&mask);
+    scalar_sub(&v, &mask);
+    scalar_compress(&v, 1);
+    scalar_encode_1(out, &v);
+}
+
+static void encode_pubkey(uint8_t *out, const ML_KEM_KEY *key)
+{
+    const uint8_t *rho = key->rho;
+    vinfo_t vinfo = key->vinfo;
+
+    vector_encode_12(out, key->t, vinfo->rank);
+    memcpy(out + vinfo->vector_bytes, rho, ML_KEM_RANDOM_BYTES);
+}
+
+static void encode_prvkey(uint8_t *out, const ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+
+    vector_encode_12(out, key->s, vinfo->rank);
+    out += vinfo->vector_bytes;
+    encode_pubkey(out, key);
+    out += vinfo->pubkey_bytes;
+    memcpy(out, key->pkhash, ML_KEM_PKHASH_BYTES);
+    out += ML_KEM_PKHASH_BYTES;
+    memcpy(out, key->z, ML_KEM_RANDOM_BYTES);
+}
+
+static int parse_pubkey(const uint8_t *in, EVP_MD_CTX *mdctx, ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+
+    /* Decode and check |t| */
+    if (!vector_decode_12(key->t, in, vinfo->rank))
+        return 0;
+    /* Save the matrix |m| recovery seed |rho| */
+    memcpy(key->rho, in + vinfo->vector_bytes, ML_KEM_RANDOM_BYTES);
+    /*
+     * Pre-compute the public key hash, needed for both encap and decap.
+     * Also pre-compute the matrix expansion, stored with the public key.
+     */
+    return hash_h(key->pkhash, in, vinfo->pubkey_bytes, mdctx, key)
+        && matrix_expand(mdctx, key);
+}
+
+/* Loading of explicit private keys is a test-only interface. */
+static int parse_prvkey(const uint8_t *in, EVP_MD_CTX *mdctx, ML_KEM_KEY *key)
+{
+    vinfo_t vinfo = key->vinfo;
+
+    /* Decode and check |s|. */
+    if (!vector_decode_12(key->s, in, vinfo->rank))
```

Comment:
still think these may be slightly better by using WPACKET and PACKET. The packet ptr then gets passed to the encode / decode sub functions.

Reply:
This function is not externally visible, the callers always provide sufficient space.
Perhaps I'll add WPACKET later, but I don't think it should hold up review at this juncture.